test/utils/test_knowledge_utils.py

```diff
@@ -0,0 +1,357 @@
+
+    entity_list = [["list"], "second"]
+    entity = apply_resolution_strategy([], entity_list, "first")
+    assert entity == ["list"]
+
+    # Test 'random' strategy
+    entity_list = ["First", "Second", "Third"]
+    for _ in range(0, 3):
+        # try multiple times
+        entity = apply_resolution_strategy([], entity_list, "first")
+        assert entity in entity_list
+
+    entity_list = ["Only"]
+    entity = apply_resolution_strategy([], entity_list, "first")
+    assert entity == "Only"
+
+    # Test 'nearest' strategy
+    # Test that no robot warns
+    entity_list = ["Only"]
+    try:
+        entity = apply_resolution_strategy([], entity_list, "nearest", None)
+    except UserWarning:
+        # test passed, this should warn
+        pass
+    else:
+        assert (
+            False
+        ), "apply_resolution_strategy didn't warn on non-existent robot strategy"
+
+    robot = Robot("test_robot")
+
+    class Entity:
+        pose = Pose()
+
+    entity_list = [Entity(), Entity(), Entity()]
+    entity_list[0].pose = Pose.from_list([1, 0])
+    entity_list[1].pose = Pose.from_list([2, 0])
+    entity_list[2].pose = Pose.from_list([3, 0])
+    entity = apply_resolution_strategy([], entity_list, "nearest", robot)
+    assert entity == entity_list[0]
+
+    robot.pose = Pose.from_list([2, 1])
+    entity = apply_resolution_strategy([], entity_list, "nearest", robot)
+    assert entity == entity_list[1]
+
+    robot.pose = Pose.from_list([1000, -1000])
+    entity = apply_resolution_strategy([], entity_list, "nearest", robot)
+    assert entity == entity_list[2]
+
+
+def test_query_to_entity():
+    test_world = load_world()
+    query = "kitchen table apple"
+    entity = query_to_entity(test_world, query.split(), "location")
+    assert entity.name == "table0_tabletop"
+    entity = query_to_entity(test_world, query.split(), "object")
+    assert entity.name == "gala"
+
+    query = "kitchen table"
+    entity = query_to_entity(test_world, query.split(), "location")
+    assert entity.name == "table0"
+
+    # if we query to a random object on the kitchen table, its parent category should be table
+    entity = query_to_entity(test_world, query.split(), "object", "random")
+    assert entity.parent.category == "table"
+
+    # nearest object on the kitchen table should be the banana
+    robot = Robot("test_robot")
+    robot.pose = Pose.from_list([1.0, -0.5])
+    query = "kitchen table"
+    entity = query_to_entity(test_world, query.split(), "object", "nearest", robot)
+    assert (
+        entity.category == "banana"
+        and entity.parent.category == "table"
+        and entity.parent.parent.parent.name == "kitchen"
+    )
+
+    # we want the nearest apple on the kitchen table, should no longer return the banana even though it is the nearest object
+    query = "kitchen table apple"
+    new_entity = query_to_entity(test_world, query.split(), "object", "nearest", robot)
+    assert new_entity != entity
```

Comment:
I'd rather have this check for the actual result rather than "not the same as the last test"